presto-main/src/main/java/io/prestosql/sql/analyzer/StatementAnalyzer.java

```diff
@@ -1243,11 +1250,24 @@ protected Scope visitTable(Table table, Optional<Scope> scope)
 
             if (isUpdateQuery) {
                 // Add the row id field
-                ColumnHandle column = metadata.getUpdateRowIdColumnHandle(session, tableHandle.get());
-                Type type = metadata.getColumnMetadata(session, tableHandle.get(), column).getType();
+                ColumnHandle rowIdColumn;
+                if (updateKind == UpdateKind.UPDATE) {
+                    Optional<List<ColumnMetadata>> updatedColumnMetadata = analysis.getUpdatedColumns();
+                    checkArgument(updatedColumnMetadata.isPresent(), "updatedColumnMetadata isn't present");
+                    Set<String> updatedColumnNames = updatedColumnMetadata.get().stream().map(ColumnMetadata::getName).collect(toImmutableSet());
+                    List<ColumnHandle> updatedColumns = columnHandles.entrySet().stream()
+                            .filter(entry -> updatedColumnNames.contains(entry.getKey()))
+                            .map(Map.Entry::getValue)
+                            .collect(toImmutableList());
+                    rowIdColumn = metadata.getUpdateRowIdColumnHandle(session, tableHandle.get(), columnHandles, updatedColumns);
+                }
+                else {
```

Comment:
Add a condition to make sure this is UpdateKind.DELETE. Otherwise, this can break in the future if a new value is added to that enum. I might structure it as:

```java
switch (updateKind) {
    case UPDATE: {
        ...
        break;
    }
    case DELETE: {
        ...
        break;
    }
    default:
        throw new UnsupportedOperationException("Update kind not supported:" + updateKind);
}
```

Reply:
Converted to a `switch` statement.